File: 34.AniTweet/app/src/main/java/com/ani/anitweet/MainActivity.java
package com.ani.anitweet;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import com.parse.LogInCallback;
import com.parse.ParseAnalytics;
import com.parse.ParseException;
import com.parse.ParseUser;
import com.parse.SignUpCallback;

import java.util.Arrays;

import io.github.inflationx.viewpump.ViewPumpContextWrapper;

public class MainActivity extends AppCompatActivity implements View.OnKeyListener, View.OnClickListener {

    ConstraintLayout backgroundLayout;
    ImageView logoImageView;
    EditText usernameEditText;
    EditText passwordEditText;

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(ViewPumpContextWrapper.wrap(newBase));
    }

    public void signupLogin(View view)
    {

        final String username = usernameEditText.getText().toString().trim();
        final String password = passwordEditText.getText().toString().trim();

        if(username.matches("") || password.matches(""))
        {
            Toast.makeText(this, "Username and Password are required", Toast.LENGTH_SHORT).show();
        }
        else
        {
            Toast.makeText(this, "Please wait...Trying to sign in", Toast.LENGTH_SHORT).show();
            ParseUser.logInInBackground(username, password, new LogInCallback() {
                @Override
                public void done(ParseUser user, ParseException e) {
                    if (user!= null)
                    {
                        Toast.makeText(MainActivity.this, "Signed in as "+user.getUsername(), Toast.LENGTH_SHORT).show();
                        showFeed();
                    }
                    else
                    {
                        new AlertDialog.Builder(MainActivity.this)
                                .setIcon(android.R.drawable.ic_menu_add)
                                .setTitle("Sign Up?")
                                .setMessage("Username or Password didn't match with existing username,\nWant to create a new user?")
                                .setPositiveButton("Create new user", new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {

                                        ParseUser user = new ParseUser();

                                        user.setUsername(username);
                                        user.setPassword(password);
                                        user.put("isFollowing", Arrays.asList());

                                        user.signUpInBackground(new SignUpCallback() {
                                            @Override
                                            public void done(ParseException e) {
                                                if(e==null)
                                                {
                                                    Toast.makeText(MainActivity.this, "Signup successful", Toast.LENGTH_SHORT).show();
                                                    showFeed();
                                                }
                                                else
                                                {
                                                    Toast.makeText(MainActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                                                }
                                            }
                                        });

                                    }
                                })
                                .setNegativeButton("Nope",null)
                                .show();
                    }
                }
            });
        }




    }

    public void showFeed()
    {
        try{
            Intent intent = new Intent( getApplicationContext(), FeedActivity.class);
            startActivity(intent);
            finish();
        }catch (Exception e)
        {
            Toast.makeText(this, e.getMessage(), Toast.LENGTH_SHORT).show();
        }

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        backgroundLayout = findViewById(R.id.backgroundLayout);
        backgroundLayout.setOnClickListener(this);
        logoImageView = findViewById(R.id.logoImageView);
        logoImageView.setOnClickListener(this);

        usernameEditText = findViewById(R.id.usernameEditText);
        passwordEditText = findViewById(R.id.passwordEditText);
        passwordEditText.setOnKeyListener(this);

        //Check if Signed In
        if(ParseUser.getCurrentUser()!=null)
        {
            Toast.makeText(this, "Signed in as "+ParseUser.getCurrentUser().getUsername(), Toast.LENGTH_SHORT).show();
            showFeed();
        }
        else
            Toast.makeText(this, "Please sign in", Toast.LENGTH_SHORT).show();


        ParseAnalytics.trackAppOpenedInBackground(getIntent());

    }

    @Override
    public void onClick(View view) {

        if(view.getId() == R.id.logoImageView || view.getId() == R.id.backgroundLayout)
        {
            InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            inputMethodManager.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
        }

    }

    @Override
    public boolean onKey(View v, int keyCode, KeyEvent event) {

        if(keyCode == KeyEvent.KEYCODE_ENTER && event.getAction() == KeyEvent.ACTION_DOWN && v.getId() == R.id.passwordEditText)
        {
            InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            inputMethodManager.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
            signupLogin(v);
        }
        return false;
    }
}

File: 35.AniSnapNChat/app/src/main/java/com/ani/anisnapnchat/MainActivity.kt
package com.ani.anisnapnchat

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.constraint.ConstraintLayout
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import io.github.inflationx.viewpump.ViewPumpContextWrapper


class MainActivity : AppCompatActivity(), View.OnClickListener, View.OnKeyListener {

    var backgroundLayout: ConstraintLayout? = null
    var logoImageView: ImageView? = null

    var emailEditText: EditText?= null
    var passwordEditText: EditText?= null

    private lateinit var auth: FirebaseAuth

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(ViewPumpContextWrapper.wrap(newBase))
    }

    fun signupLogin(view: View)
    {
        val email = emailEditText!!.text.toString().trim()
        val password = passwordEditText!!.text.toString().trim()

        if (email.matches("^$".toRegex()) || password.matches("^$".toRegex()))
            Toast.makeText(this@MainActivity, "Email or Password can't be blank",Toast.LENGTH_SHORT).show()

        //else if (!email.matches("^.*@.*.\\..+$".toRegex())) //OR
        else if (!email.matches(android.util.Patterns.EMAIL_ADDRESS.toRegex()))
            Toast.makeText(this@MainActivity, "Invalid Email",Toast.LENGTH_SHORT).show()

        else {

            //Try to sign in
            auth.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(this) { task ->
                    if (task.isSuccessful) {
                        // Sign in success
                        logIn()
                    } else {

                        // Sign in fails, ask for signup

                        AlertDialog.Builder(this@MainActivity)
                            .setIcon(android.R.drawable.ic_menu_add)
                            .setTitle("Sign Up?")
                            .setMessage("Username or Password didn't match with existing username,\nWant to create a new user?")
                            .setPositiveButton("Create new user") { dialog, which ->

                                auth.createUserWithEmailAndPassword(email, password)
                                    .addOnCompleteListener(this) { task ->
                                        if (task.isSuccessful) {
                                            // Sign up success
                                            Toast.makeText(this@MainActivity, "Signup successful", Toast.LENGTH_SHORT).show()

                                            //Add user to Database
                                            FirebaseDatabase.getInstance().getReference().child("Users").child(task.result!!.user.uid).child("email").setValue(email)

                                            logIn()
                                        } else {
                                            // Sign Up fails
                                            Toast.makeText(baseContext, task.exception!!.message, Toast.LENGTH_SHORT).show()
                                        }

                                        // ...
                                    }

                            }
                            .setNegativeButton("Nope", null)
                            .show()

                    }

                }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        emailEditText = findViewById(R.id.emailEditText)
        passwordEditText = findViewById(R.id.passwordEditText)

        backgroundLayout = findViewById(R.id.backgroundLayout);
        backgroundLayout!!.setOnClickListener(this);
        logoImageView = findViewById(R.id.logoImageView);
        logoImageView!!.setOnClickListener(this);
        passwordEditText!!.setOnKeyListener(this);

        // Initialize Firebase Auth
        auth = FirebaseAuth.getInstance()

        if(auth.currentUser != null)
        {
            logIn()
        }
    }

    fun logIn()
    {
        Toast.makeText(this@MainActivity,"Signed in as "+auth.currentUser!!.email, Toast.LENGTH_SHORT).show()

        //Move to next activity
        val intent = Intent(this, SnapsActivity::class.java)
        startActivity(intent)
        finish()
    }

    override fun onClick(view: View) {

        if (view.id == R.id.logoImageView || view.id == R.id.backgroundLayout) {
            val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
        }

    }

    override fun onKey(v: View, keyCode: Int, event: KeyEvent): Boolean {

        if (keyCode == KeyEvent.KEYCODE_ENTER && event.action == KeyEvent.ACTION_DOWN && v.id == R.id.passwordEditText) {
            val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
            signupLogin(v)
            return true
        }
        else
            return false
    }

}
